File: backend/app/schemas/ticket_processing.py
"""API response contracts for completed ticket-processing runs."""

from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.agents.classification_agent import IntentLabel, SentimentLabel
from app.agents.guardrail_models import GuardrailRule
from app.agents.planning_models import PlanningAction
from app.agents.priority_agent import PriorityLevel
from app.agents.resolution_models import ResolutionOutcome
from app.models.ticket import TicketPriority


class ClassificationResultResponse(BaseModel):
    """Classification output returned by the completed pipeline."""

    sentiment: SentimentLabel
    sentiment_score: float
    sentiment_confidence: float
    intent: IntentLabel
    intent_confidence: float
    used_fallback: bool


class PriorityResultResponse(BaseModel):
    """Priority output returned by the completed pipeline."""

    level: PriorityLevel
    score: int
    ticket_priority: TicketPriority


class GuardrailRuleResponse(BaseModel):
    """One deterministic guardrail rule evaluation."""

    rule: GuardrailRule
    passed: bool
    reason: Optional[str]
    risk_points: int


class GuardrailResultResponse(BaseModel):
    """Aggregated guardrail result returned by the pipeline."""

    passed: bool
    risk_score: int
    failure_reason: Optional[str]
    requires_human: bool
    evaluated_rules: list[GuardrailRuleResponse]


class ResolutionOutcomeResponse(BaseModel):
    """Final resolution output and its persisted record identifier."""

    outcome: ResolutionOutcome
    message: str
    performed_action: Optional[PlanningAction]
    requires_follow_up: bool
    follow_up_reason: Optional[str]
    persisted_resolution_id: UUID


class TicketProcessingResponse(BaseModel):
    """Structured API result for one completed ticket-processing execution."""

    ticket_id: UUID
    correlation_id: Optional[str] = None
    classification: ClassificationResultResponse
    priority: PriorityResultResponse
    planning_action: PlanningAction
    planning_requires_human: bool
    confidence: float = Field(ge=0.0, le=1.0)
    guardrail: GuardrailResultResponse
    resolution: ResolutionOutcomeResponse
    requires_human: bool

File: frontend/models/ticket-processing.ts
import type { TicketPriority } from "@/models/ticket";

export type SentimentLabel = "positive" | "neutral" | "negative";
export type IntentLabel =
  | "account_access"
  | "billing"
  | "cancellation"
  | "delivery"
  | "product_issue"
  | "refund"
  | "general_support";
export type PriorityLevel = "critical" | "high" | "standard";
export type PlanningAction =
  | "refund"
  | "escalate"
  | "troubleshoot"
  | "ask_for_information"
  | "cancel_order"
  | "update_account";
export type GuardrailRule =
  | "required_context"
  | "refund_approval_limit"
  | "refund_eligibility_window"
  | "account_flagged"
  | "excessive_recent_refunds"
  | "high_risk_customer_state";
export type ResolutionOutcome = "automated_response" | "human_review" | "information_requested";

export type ClassificationResultResponse = {
  sentiment: SentimentLabel;
  sentiment_score: number;
  sentiment_confidence: number;
  intent: IntentLabel;
  intent_confidence: number;
  used_fallback: boolean;
};

export type PriorityResultResponse = {
  level: PriorityLevel;
  score: number;
  ticket_priority: TicketPriority;
};

export type GuardrailRuleResponse = {
  rule: GuardrailRule;
  passed: boolean;
  reason: string | null;
  risk_points: number;
};

export type GuardrailResultResponse = {
  passed: boolean;
  risk_score: number;
  failure_reason: string | null;
  requires_human: boolean;
  evaluated_rules: GuardrailRuleResponse[];
};

export type ResolutionOutcomeResponse = {
  outcome: ResolutionOutcome;
  message: string;
  performed_action: PlanningAction | null;
  requires_follow_up: boolean;
  follow_up_reason: string | null;
  persisted_resolution_id: string;
};

export type TicketProcessingResponse = {
  ticket_id: string;
  correlation_id: string | null;
  classification: ClassificationResultResponse;
  priority: PriorityResultResponse;
  planning_action: PlanningAction;
  planning_requires_human: boolean;
  confidence: number;
  guardrail: GuardrailResultResponse;
  resolution: ResolutionOutcomeResponse;
  requires_human: boolean;
};
